src: include sync stop for heap object tracking

juanarbol opened the pull request:
This patch includes `StopTrackingHeapObjects` synchronously. NSolid will
grab all in-progress profiles on exit.

Also, this path will convert the tracker stopper serialization async.

PR-URL: https://github.com/nodesource/nsolid/pull/92

juanarbol committed:
src: include sync stop for heap object tracking

This patch includes `StopTrackingHeapObjects` synchronously. NSolid will
grab all in-progress profiles on exit.

Also, this path will convert the tracker stopper serialization async.

Signed-off-by: Juan José Arboleda <soyjuanarbol@gmail.com>
PR-URL: https://github.com/nodesource/nsolid/pull/92
Reviewed-by: Trevor Norris <trev.norris@gmail.com>
Reviewed-by: Santiago Gimeno <santiago.gimeno@gmail.com>

santigimeno commented on src/nsolid.h:
```diff
@@ -1005,6 +1005,16 @@ class NODE_EXTERN Snapshot {
    */
   static int StopTrackingHeapObjects(SharedEnvInst envinst);
 
+  /**
+   * @brief same as TakeSnapshot but stops tracking heap objects in the heap
+   * profiler synchronously.
```
Seems TakeSnapshot is a leftover

santigimeno commented on src/nsolid/nsolid_heap_snapshot.cc:
```diff
@@ -83,6 +83,71 @@ int NSolidHeapSnapshot::StopTrackingHeapObjects(SharedEnvInst envinst) {
+            std::string(),
+            stor.data.get());
+  } else {
+    std::string snapshot_str;
+    DataOutputStream<uint64_t, v8::HeapSnapshot> stream(&snapshot_str,
+                                                        snapshot,
+                                                        &thread_id);
+
```
Why can't we use the "streaming" version of the DataOutputStream here. Both are sync so we should be fine

juanarbol replied:
This comment as discussed is for the async stoped. Fixed in https://github.com/nodesource/nsolid/commit/37d63c035e309c433c4af1c02774d4aec1f79ee5

santigimeno commented on src/nsolid/nsolid_heap_snapshot.cc:
```diff
@@ -83,6 +83,71 @@ int NSolidHeapSnapshot::StopTrackingHeapObjects(SharedEnvInst envinst) {
+    // Tell ZMQ that the snapshot is done
+    stor.cb(0, std::string(), stor.data.get());
+
+    // Don't leak the snapshot string
+    snapshot_str.clear();
```
I don't understand why we're doing this

santigimeno reviewed (changes requested):
Looks good: I'll test in a few minutes

santigimeno reviewed (changes requested):
Just tested, everything's good. I'll approve once the nits are addressed

santigimeno committed:
src,test: multiple fixes in heap profile code

Better handling of state in `HeapSnapshotStor` to avoid possible crashes
and race conditions.
Make the test reliable by adding a JS callback to notify Heap Profile
completions.

PR-URL: https://github.com/nodesource/nsolid/pull/92
Reviewed-by: Trevor Norris <trev.norris@gmail.com>

trevnorris reviewed (approved):
Everything now passes with the additional changes.
